lib/spark.js

```diff
@@ -0,0 +1,107 @@
+import { assertEquals } from 'zinnia:assert'
+
+const sleep = dt => new Promise(resolve => setTimeout(resolve, dt))
+
+export default class Spark {
+  #fetch
+
+  constructor ({ fetch = globalThis.fetch } = {}) {
+    this.#fetch = fetch
+  }
+
+  async getRetrieval () {
+    console.log('Geting retrieval...')
+    const res = await this.#fetch('https://spark.fly.dev/retrievals', {
+      method: 'POST'
+    })
+    const retrieval = await res.json()
+    console.log({ retrieval })
+    return retrieval
+  }
+  
+  async fetchCAR (url) {
+    // TODO: Ensure these fields are persisted by `spark-api`
+    const stats = {
+      start: new Date(),
+      firstByte: null,
+      end: null,
+      byteLength: 0,
+      status: null
+    }
+    console.log('Fetching CAR...')
+    const controller = new AbortController()
+    const { signal } = controller
+    const res = await this.#fetch(url, { signal })
+    stats.status = res.status
+  
+    if (res.ok) {
+      // Abort if no data is received for 10 seconds
+      let timeout
+      const startTimeout = () => {
+        if (timeout) {
+          clearTimeout(timeout)
+        }
+        timeout = setTimeout(() => controller.abort(), 10_000)
+      }
+      startTimeout()
+  
+      try {
+        for await (const value of res.body) {
+          if (stats.firstByte === null) {
+            stats.firstByte = new Date()
+          }
+          stats.byteLength += value.byteLength
+          startTimeout()
+        }
+      } finally {
+        clearTimeout(timeout)
+      }
+    }
+    stats.end = new Date()
+    console.log(stats)
+    return stats
+  }
+  
+  async submitRetrieval (id, stats) {
+    console.log('Submitting retrieval...')
+    const res = await this.#fetch(`https://spark.fly.dev/retrievals/${id}`, {
+      method: 'PATCH',
+      body: JSON.stringify(stats),
+      headers: {
+        'Content-Type': 'application/json'
+      }
+    })
+    assertEquals(res.status, 200)
+    console.log('Retrieval submitted')
+  }
+
+  async nextRetrieval () {
+    const retrieval = await this.getRetrieval()
+
+    let success = false
+    let stats
+    const url = `https://strn.pl/ipfs/${retrieval.cid}`
+    try {
+      stats = await this.fetchCAR(url)
+      success = true
+    } catch (err) {
+      console.error(`Failed to fetch ${url}`)
+      console.error(err)
+    }
+
+    await this.submitRetrieval(retrieval.id, { success, ...stats })
+    Zinnia.jobCompleted()
+  }
+
+  async run () {
+    while (true) {
+      try {
+        await this.nextRetrieval()
+      } catch (err) {
+        Zinnia.activity.error('SPARK failed reporting retrieval')
```

Comment:
> can you please suggest how to test it?

Hmm, this is a great question!

We need to intercept `Zinnia.activity.*` calls to verify whether an activity was logged or not.

Maybe we can inject `Zinnia.activity` object into the `ActivityState` class, using `Zinnia.activity` as the default value?

Then we can write tests as follows, using hand-written spies or Sinon.js:

```js
test('first error is reported', () => {
  const success = sinon.spy();
  const error = sinon.spy();
  const state = new ActivityState({ success, error });
  
  state.onError();
  
  sinon.assertNotCalled(success);
  sinon.assertCalled(error);
});

// similarly for other test cases:
// - onError(), onError()
// - onSuccess()
// - onSuccess(), onSuccess()
// - onError(), onSuccess()
```

I expect that many Zinnia/Station modules will need to aggregate these error/success activities. Maybe this is something we should implement in Zinnia (both CLI and Daemon). I am thinking about the following design to begin with, we can iterate on it later:
- The activity reporter remembers the last activity reported by the module.
- If the module reports the same activity, then this duplicate activity is ignored.
- The modules can report activity for every success/error, as you are doing now. 

I see two downsides:
- The "success" activity message must be generic to cover both the first successful retrieval after starting and the first successful retrieval after being in an error state.
- At Station startup, we will get one activity entry for each module when they complete their first job. I guess this can actually be considered a nice feature, as users will know which modules are working well and which are not. At the cost of having a slightly noisy startup.

WDYT?

Reply:
Diffing Activity strings sounds brittle to me 🤔  There can also be different types of errors, and the module can be considered online / healthy even if it reports that a certain activity failed. I think therefore we should watch the user space some more, to discover the patterns first.

If you look at the Activity class, it's 15 SLOC with trivial logic. Do you think it's worth adding tests for this if it involves importing a spy library, and adding more dependency injection? I personally wouldn't test this class, at this stage of the project (walking skeleton).

Reply:
> Diffing Activity strings sounds brittle to me 🤔 There can also be different types of errors, and the module can be considered online / healthy even if it reports that a certain activity failed. I think therefore we should watch the user space some more, to discover the patterns first.

Makes sense 👍🏻 
 
> If you look at the Activity class, it's 15 SLOC with trivial logic. Do you think it's worth adding tests for this if it involves importing a spy library, and adding more dependency injection? I personally wouldn't test this class, at this stage of the project (walking skeleton).

The goal of the Walking Skeleton is to implement the necessary infrastructure to make it easy to add more features and tests later in the project. If we already have several places needed mocks, then yes, it's time to add the library now! In my mind, Walking Skeleton is about discovering the different architectural components we need in the project and paying the cost of integration now, so that we can move faster in the future.

Having said that, I agree with you this class is so simple we don't need to test it yet.

Reply:
I disagree on whether we have identified that this projects needs spys and mocks. Let's sync on this in person please 👍